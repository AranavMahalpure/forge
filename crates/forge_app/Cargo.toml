[package]
name = "forge_app"
version = "0.1.0"
edition = "2024"

[dependencies]
anyhow = "1.0.95"
async-trait = "0.1.85"
forge_domain = { path = "../forge_domain" }
tokio-stream = "0.1.17"
