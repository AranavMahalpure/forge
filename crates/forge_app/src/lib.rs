mod ide;
mod repo;
mod schema;
mod service;
mod sqlite;

pub use service::{APIService, Service};
