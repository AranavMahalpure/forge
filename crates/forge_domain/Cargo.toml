[package]
name = "forge_domain"
version = "0.1.0"
edition = "2024"

[dependencies]
derive_more = { version = "1.0.0-beta.6", features = ["from", "error", "display"] }
thiserror = "1.0.56"
url = "2.5.4"
