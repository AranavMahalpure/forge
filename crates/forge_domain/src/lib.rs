mod agent;
mod chat_request;
mod chat_response;
mod chat_stream_ext;
mod command;
mod config;
mod context;
mod conversation;
mod embedding;
mod environment;
mod errata;
mod error;
mod learning;
mod message;
mod model;
mod provider;
mod stream_ext;
mod tool;
mod tool_call;
mod tool_call_parser;
mod tool_call_service;
mod tool_choice;
mod tool_definition;
mod tool_name;
mod tool_result;
mod tool_usage;
mod user_interaction;

pub use agent::*;
pub use chat_request::*;
pub use chat_response::*;
pub use chat_stream_ext::*;
pub use command::*;
pub use config::*;
pub use context::*;
pub use conversation::*;
pub use embedding::*;
pub use environment::*;
pub use errata::*;
pub use error::*;
pub use learning::*;
pub use message::*;
pub use model::*;
pub use provider::*;
pub use stream_ext::*;
pub use tool::*;
pub use tool_call::*;
pub use tool_call_parser::*;
pub use tool_call_service::*;
pub use tool_choice::*;
pub use tool_definition::*;
pub use tool_name::*;
pub use tool_result::*;
pub use tool_usage::*;
pub use user_interaction::*;
