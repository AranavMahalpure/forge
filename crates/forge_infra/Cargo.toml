[package]
name = "forge_infra"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0.95"
async-trait = "0.1.86"
dirs = "6.0.0"
dotenv = "0.15.0"
forge_domain = { path = "../forge_domain" }
forge_app = { path = "../forge_app" }
tokio = "1.43.0"
diesel = { version = "2.2.7", features = ["chrono", "r2d2"] }
tokio-retry = "0.3.0"
derive_setters = "0.1.6"
chrono = { version = "0.4.39" }
uuid = "1.13.1"
serde_json = "1.0.138"
