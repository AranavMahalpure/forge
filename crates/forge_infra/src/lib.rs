mod conn;
mod env;
mod file_read;
mod infra;
mod knowledge;

pub use infra::*;
