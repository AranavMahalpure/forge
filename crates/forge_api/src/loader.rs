use std::path::Path;
use std::sync::Arc;

use forge_app::{FileReadService, Infrastructure};
use forge_domain::Workflow;

/// A workflow loader to load the workflow from the given path.
/// It also resolves the internal paths specified in the workflow.
pub struct ForgeLoaderService<F>(Arc<F>);

impl<F> ForgeLoaderService<F> {
    pub fn new(app: Arc<F>) -> Self {
        Self(app)
    }
}

impl<F: Infrastructure> ForgeLoaderService<F> {
    /// loads the workflow from the given path.
    pub async fn load(&self, path: &Path) -> anyhow::Result<Workflow> {
        let workflow_content = self.0.file_read_service().read(path).await?;
        let workflow: Workflow = workflow_content.parse()?;

        Ok(workflow)
    }
}

#[cfg(test)]
mod tests {
    use std::path::{Path, PathBuf};
    use std::sync::Arc;

    use anyhow::Result;
    use forge_domain::Workflow;
    use forge_infra::ForgeInfra;
    use tempfile::TempDir;

    use super::ForgeLoaderService;

    const BASE_WORKFLOW: &str = r#"
[[agents]]
id = "developer"
model = "anthropic/claude-3.5-sonnet"
tools = ["tool_forge_fs_read", "tool_forge_fs_create"]
subscribe = ["user_task"]
max_turns = 1024"#;

    struct Fixture {
        temp_dir: TempDir,
        workflow_path: PathBuf,
        loader: ForgeLoaderService<ForgeInfra>,
    }

    impl Default for Fixture {
        fn default() -> Self {
            let temp_dir = tempfile::tempdir().unwrap();
            let loader = ForgeLoaderService::new(Arc::new(ForgeInfra::new(true)));
            Self {
                temp_dir,
                loader,
                workflow_path: PathBuf::from("workflow.toml"),
            }
        }
    }

    impl Fixture {
        async fn run(self, system_prompt: &str, user_prompt: &str) -> Result<Workflow> {
            let workflow = format!(
                "{}\n\n[agents.system_prompt]\ntemplate = \"{}\"\n\n[agents.user_prompt]\ntemplate = \"{}\"",
                BASE_WORKFLOW, system_prompt, user_prompt
            );
            let workflow_path = self.temp_dir.path().join(self.workflow_path.clone());
            tokio::fs::write(&workflow_path, workflow).await?;
            self.loader
                .load(&self.temp_dir.path().join(self.workflow_path.clone()))
                .await
        }

        async fn create_prompt_file(&self, path: &Path, content: &str) -> Result<()> {
            let file_path = self.temp_dir.path().join(path);
            if let Some(parent) = file_path.parent() {
                tokio::fs::create_dir_all(parent).await?;
            }
            tokio::fs::write(file_path, content).await?;
            Ok(())
        }
    }

    #[tokio::test]
    async fn test_load_workflow_with_string_literals() -> Result<()> {
        let workflow = Fixture::default()
            .run(
                "You are a software developer assistant",
                "<task>{{event.value}}</task>",
            )
            .await?;
        insta::assert_snapshot!(serde_json::to_string_pretty(&workflow)?);
        Ok(())
    }

    #[tokio::test]
    async fn test_load_workflow_with_file_prompt() -> Result<()> {
        let fixture = Fixture::default();
        // create system prompt file
        let system_prompt = "You are a software developer assistant, good at solving complex software engineering problems";
        let system_prompt_path = Path::new("nested/system_prompt.md");
        fixture
            .create_prompt_file(system_prompt_path, system_prompt)
            .await?;

        let workflow = fixture
            .run("nested/system_prompt.md", "<task>{{event.value}}</task>")
            .await?;
        insta::assert_snapshot!(serde_json::to_string_pretty(&workflow)?);
        Ok(())
    }

    #[tokio::test]
    async fn test_load_workflow_with_missing_prompt_file() -> Result<()> {
        let workflow = Fixture::default()
            .run("nested/system_prompt.md", "<task>{{event.value}}</task>")
            .await?;
        insta::assert_snapshot!(serde_json::to_string_pretty(&workflow)?);
        Ok(())
    }
}
