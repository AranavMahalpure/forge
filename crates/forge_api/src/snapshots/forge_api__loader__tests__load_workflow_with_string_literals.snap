---
source: crates/forge_api/src/loader.rs
expression: "serde_json::to_string_pretty(&workflow)?"
snapshot_kind: text
---
{
  "agents": [
    {
      "id": "developer",
      "model": "anthropic/claude-3.5-sonnet",
      "description": null,
      "system_prompt": {
        "template": "You are a software developer assistant"
      },
      "user_prompt": {
        "template": "<task>{{event.value}}</task>"
      },
      "tools": [
        "tool_forge_fs_read",
        "tool_forge_fs_create"
      ],
      "subscribe": [
        "user_task"
      ],
      "max_turns": 1024
    }
  ]
}
